lib/common/util/exceptions.js

```diff
@@ -0,0 +1,130 @@
+// @flow
+
+import moment from 'moment'
+
+import type {ParsedRangesAndDates} from '../../types'
+
+// Method to filter out new dates (0's) so that flow is happy
+// We have to use a reducer to keep flow happy.
+export function filterOutNewDates (dates: Array<string | 0>): Array<string> {
+  const outputDates: Array<string> = []
+  const filter = (acc, current) => {
+    typeof current === 'string' && acc.push(current)
+    return acc
+  }
+  dates.reduce(filter, outputDates)
+  return outputDates
+}
+
+/**
+ * This method adds a range of dates from startRange to endRange to the dates array, inclusively according to startInclusive
+ * and endInclusive.
+ */
+export function addRangeOfDates (
+  startRange: moment,
+  endRange: moment,
+  dates: Array<string>,
+  startInclusive: boolean,
+  endInclusive: boolean
+) {
+  const datesToAdd = []
+  while (endRange.unix() >= startRange.unix()) {
+    if (endRange.unix() === startRange.unix() && !endInclusive) break
+    if (datesToAdd.length === 0 && !startInclusive) {
+      startInclusive = true
+      continue
+    }
+    datesToAdd.push(startRange.format('YYYYMMDD'))
+    startRange.add(1, 'days')
+  }
+  datesToAdd.forEach(date => {
+    if (!dates.some(el => el === date)) {
+      dates.push(date)
+    }
+  })
+  return dates
+}
+
+/**
+ * This method removes a range of dates from startRange to endRange to the dates array, inclusively according to startInclusive
+ * and endInclusive.
+ */
+export function removeRangeOfDates (
+  startRange: moment,
+  endRange: moment,
+  dates: Array<string>,
+  startInclusive: boolean,
+  endInclusive: boolean
+) {
+  const datesToRemove = []
+  while (endRange.unix() >= startRange.unix()) {
+    if (endRange.unix() === startRange.unix() && !endInclusive) break
+    if (datesToRemove.length === 0 && !startInclusive) {
+      startInclusive = true
+      startRange.add(1, 'days')
+      continue
+    }
+    datesToRemove.push(startRange.format('YYYYMMDD'))
+    startRange.add(1, 'days')
+  }
+  dates = (dates.filter(date => !datesToRemove.some(el => el === date)): Array<string>)
+  return dates
+}
+
+function datesAreConsecutive (next: ?string, currentDate: moment): boolean {
+  if (!next) return false
+  const nextDate = moment(next)
+
+  if (next && nextDate.diff(currentDate, 'days') === 1) return true
+  else return false
+}
+
+function previousDateIsConsecutive (prev: ?string, currentDate: moment): boolean {
+  if (!prev) return false
+  const prevDate = moment(prev)
+
+  if (prev && currentDate.diff(prevDate, 'days') === 1) return true
+  else return false
+}
+
+/**
+ * This method extracts any ranges from a list of dates that are either being generated
+ * or obtained from the back end.
+ */
+export function createDateRangesFromDates (acc: ParsedRangesAndDates, current: string, currentIndex: number, dates: Array<string>) {
```

Comment:
If we're going to have a method this complicated, then we need tests for it. 

I personally would be in favor of again using recursion to clean this up. With one list of range starts, one list of range ends, and a list of dates, the dates array can be processed without needing `current` or `currentIndex`.

Reply:
This method was cleaned up a lot in https://github.com/ibi-group/datatools-ui/pull/828/commits/0bd970887b85994ed9f9452ec5640b01eb021f28!